src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -461,7 +463,15 @@ bool IsDerivedType(TypeSymbol possibleDerived, TypeSymbol possibleBase)
                 var tests = ArrayBuilder<Tests>.GetInstance(2);
                 var convertedInput = MakeConvertToType(input, constant.Syntax, constant.Value.Type!, isExplicitTest: false, tests);
                 output = convertedInput;
-                tests.Add(new Tests.One(new BoundDagValueTest(constant.Syntax, constant.ConstantValue, convertedInput)));
+                if (ValueSetFactory.ForInput(input)?.Related(BinaryOperatorKind.Equal, constant.ConstantValue).IsEmpty == true)
+                {
```

Comment:
Could we add an assertion that we're indeed dealing with a length input (as the comment explains)?